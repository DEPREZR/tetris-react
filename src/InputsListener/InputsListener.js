import React, { useState } from 'react';
import Hotkeys from 'react-hot-keys';
import PropTypes from 'prop-types';

export const InputsContext = React.createContext();

export const handleKeyPressed = ({ pressed, setPressed }) => () =>
  setPressed(pressed);

const InputsListener = ({ children }) => {
  const [pressedDown, setPressedDown] = useState(false);
  const [pressedUp, setPressedUp] = useState(false);
  const [pressedRight, setPressedRight] = useState(false);
  const [pressedLeft, setPressedLeft] = useState(false);

  return (
    <React.Fragment>
      <Hotkeys
        keyName="down"
        onKeyDown={handleKeyPressed({
          pressed: true,
          setPressed: setPressedDown
        })}
        onKeyUp={handleKeyPressed({
          pressed: false,
          setPressed: setPressedDown
        })}
      />
      <Hotkeys
        keyName="up"
        onKeyDown={handleKeyPressed({
          pressed: true,
          setPressed: setPressedUp
        })}
        onKeyUp={handleKeyPressed({
          pressed: false,
          setPressed: setPressedUp
        })}
      />
      <Hotkeys
        keyName="right"
        onKeyDown={handleKeyPressed({
          pressed: true,
          setPressed: setPressedRight
        })}
        onKeyUp={handleKeyPressed({
          pressed: false,
          setPressed: setPressedRight
        })}
      />
      <Hotkeys
        keyName="left"
        onKeyDown={handleKeyPressed({
          pressed: true,
          setPressed: setPressedLeft
        })}
        onKeyUp={handleKeyPressed({
          pressed: false,
          setPressed: setPressedLeft
        })}
      />
      <InputsContext.Provider
        value={{ pressedDown, pressedUp, pressedRight, pressedLeft }}
      >
        {children}
      </InputsContext.Provider>
    </React.Fragment>
  );
};

InputsListener.propTypes = {
  children: PropTypes.node
};

export default InputsListener;
