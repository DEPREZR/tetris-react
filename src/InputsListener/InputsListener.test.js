import React from 'react';
import { shallow } from 'enzyme';
import InputsListener, { handleKeyPressed } from '.';

describe('InputsListener', () => {
  it('renders correctly', () => {
    const wrapper = shallow(
      <InputsListener>
        <div>test</div>
      </InputsListener>
    );

    expect(wrapper).toMatchSnapshot();
  });
});

describe('handleKeyPressed', () => {
  it('should setPressed with pressed and call e.preventDefault()', () => {
    const params = { pressed: true, setPressed: jest.fn() };
    const e = { preventDefault: jest.fn() };

    handleKeyPressed(params)(null, e);

    expect(params.setPressed).toHaveBeenCalledWith(params.pressed);
    expect(e.preventDefault).toHaveBeenCalledWith();

    params.pressed = false;
    e.preventDefault.mockClear();
    params.setPressed.mockClear();

    handleKeyPressed(params)(null, e);

    expect(params.setPressed).toHaveBeenCalledWith(params.pressed);
    expect(e.preventDefault).toHaveBeenCalledWith();
  });
});
