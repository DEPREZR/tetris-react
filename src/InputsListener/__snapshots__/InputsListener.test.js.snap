// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputsListener renders correctly 1`] = `
<Fragment>
  <ReactHotkeys
    keyName="down"
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  />
  <ReactHotkeys
    keyName="up"
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  />
  <ReactHotkeys
    keyName="right"
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  />
  <ReactHotkeys
    keyName="left"
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  />
  <ContextProvider
    value={
      Object {
        "pressedDown": false,
        "pressedLeft": false,
        "pressedRight": false,
        "pressedUp": false,
      }
    }
  >
    <div>
      test
    </div>
  </ContextProvider>
</Fragment>
`;
